File: samples/java/src/app/GraphsPrimitivesEdges.java
package app;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class GraphsPrimitivesEdges {

    public static void main(String[] args) {
        // graph with hashmap primitives
        HashMap<String, List<List<String>>> gmap = new HashMap<>();
        gmap.put("1", new ArrayList<>(List.of(List.of("2", "3"), List.of("10",
                "20"))));
        gmap.put("2", new ArrayList<>(List.of(List.of("3", "4"), List.of("30",
                "40"))));
        gmap.put("3", new ArrayList<>());
        gmap.put("4", new ArrayList<>());
        gmap.put("5", new ArrayList<>());

        while (true) {
        }
    }

    // custom extractor
    static class Extractor {
        // register the custom types
        public static List<String> registerTypes() {
            return new ArrayList<>(List.of("HashMap", "HashMap$Entry",
                    "HashMap$Node"));
        }

        public static List<?> getNodes(
                Object node,
                HashMap<String, List<List<String>>> root) {
            if (node instanceof HashMap) {
                List<HashMap.Entry<String, List<List<String>>>> nodes = new ArrayList<>();
                nodes.add(root.entrySet().iterator().next());
                return nodes;
            } else if (node instanceof HashMap.Entry) {
                List<HashMap.Entry<String, List<List<String>>>> nodes = new ArrayList<>();
                HashMap.Entry<String, List<List<String>>> nodeObj = (HashMap.Entry<String, List<List<String>>>) node;
                if(nodeObj.getValue().size() == 0) {
                    return nodes;
                }
                HashSet<String> keys = new HashSet<>(nodeObj.getValue().get(0));
                for (HashMap.Entry<String, List<List<String>>> entry : root.entrySet()) {
                    if (keys.contains(entry.getKey())) {
                        nodes.add(entry);
                    }
                }
                return nodes;
            }
            return new ArrayList<>();
        }

        public static List<?> getEdges(
                Object node,
                HashMap<String, List<List<String>>> root) {
            if (node instanceof HashMap) {
                return new ArrayList<>(List.of(""));
            } else if (node instanceof HashMap.Entry) {
                HashMap.Entry<String, List<List<String>>> nodeObj = (HashMap.Entry<String, List<List<String>>>) node;
                return nodeObj.getValue().get(1);
            }
            return new ArrayList<>();
        }

        public static String toString(Object node) {
            if(node instanceof HashMap) {
                return "";
            } else if(node instanceof HashMap.Entry) {
                return String.valueOf(((HashMap.Entry) node).getKey());
            }
            return "";
        }
    }
}

File: samples/csharp/src/GraphsPrimitivesEdges.cs
using System.Collections.Generic;
using System.Linq;

public class GraphsPrimitivesEdges
{

    public static void RunMain(string[] args)
    {
        // graph with hashmap primitives
        Dictionary<string, List<List<string>>> gmap
            = new Dictionary<string, List<List<string>>>();
        gmap["1"] = new List<List<string>>([["2", "3"], ["10", "20"]]);
        gmap["2"] = new List<List<string>>([["3", "4"], ["30", "40"]]);
        gmap["3"] = new List<List<string>>();
        gmap["4"] = new List<List<string>>();
        gmap["5"] = new List<List<string>>();
    }

    // custom extractor
    static class Extractor
    {
        // register the custom types
        public static List<string> RegisterTypes()
        {
            return new List<string>(["*"]);
        }

        public static List<KeyValuePair<string, List<List<string>>>> GetNodes(
                object node,
                Dictionary<string, List<List<string>>> root)
        {
            if (node.GetType() == typeof(Dictionary<string, List<List<string>>>))
            {
                List<KeyValuePair<string, List<List<string>>>> nodes = [root.First()];
                return nodes;
            }
            else if (node.GetType() == typeof(KeyValuePair<string, List<List<string>>>))
            {
                List<KeyValuePair<string, List<List<string>>>> nodes
                    = new List<KeyValuePair<string, List<List<string>>>>();
                KeyValuePair<string, List<List<string>>> nodeObj
                    = (KeyValuePair<string, List<List<string>>>)node;
                if (nodeObj.Value.Count == 0)
                {
                    return nodes;
                }
                HashSet<string> keys = new HashSet<string>(nodeObj.Value[0]);
                foreach (KeyValuePair<string, List<List<string>>> entry in root)
                {
                    if (keys.Contains(entry.Key))
                    {
                        nodes.Add(entry);
                    }
                }
                return nodes;
            }
            return new List<KeyValuePair<string, List<List<string>>>>();
        }

        public static List<string> GetEdges(
                object node,
                Dictionary<string, List<List<string>>> root)
        {
            if (node.GetType() == typeof(Dictionary<string, List<List<string>>>))
            {
                return new List<string>([""]);
            }
            else if (node.GetType() == typeof(KeyValuePair<string, List<List<string>>>))
            {
                KeyValuePair<string, List<List<string>>> nodeObj
                    = (KeyValuePair<string, List<List<string>>>)node;
                List<string> edges = new List<string>();
                edges.AddRange(nodeObj.Value[1]);
                return edges;
            }
            return new List<string>();
        }

        public static string ToString(object node)
        {
            if (node.GetType() == typeof(Dictionary<string, List<List<string>>>))
            {
                return "";
            }
            if (node.GetType() == typeof(KeyValuePair<string, List<List<string>>>))
            {
                return ((KeyValuePair<string, List<List<string>>>)node).Key.ToString();
            }
            return "";
        }
    }
}